src/memory/mmap_bytebuf.rs

```diff
@@ -0,0 +1,199 @@
+            // - `self.len` cannot be larger than `isize::max_value()`.
+            slice::from_raw_parts(self.ptr.as_ptr(), self.len)
+        }
+    }
+
+    fn as_slice_mut(&mut self) -> &mut [u8] {
+        unsafe {
+            // Safety Proof:
+            // - See the proof for `Self::as_slice`
+            // - Additionally, it is not possible to obtain two mutable references for `self.ptr`
+            slice::from_raw_parts_mut(self.ptr.as_ptr(), self.len)
+        }
+    }
+}
+
+impl Drop for Mmap {
+    fn drop(&mut self) {
+        let ret_val = unsafe {
+            // Safety proof:
+            // - `self.ptr` was allocated by a call to `mmap`.
+            // - `self.len` was saved at the same time and it doesn't change throughout the lifetime
+            //   of `self`.
+            libc::munmap(self.ptr.as_ptr() as *mut libc::c_void, self.len)
+        };
+
+        // There is no reason for `munmap` to fail to deallocate a private annonymous mapping
+        // allocated by `mmap`.
+        // However, for the cases when it actually fails prefer to fail, in order to not leak
+        // and exhaust the virtual memory.
+        assert_eq!(ret_val, 0, "munmap failed");
+    }
+}
+
+pub struct ByteBuf {
+    mmap: Option<Mmap>,
+}
+
+impl ByteBuf {
+    pub fn new(len: usize) -> Result<Self, &'static str> {
+        let mmap = if len == 0 {
+            None
+        } else {
+            Some(Mmap::new(len)?)
+        };
+        Ok(Self { mmap })
+    }
+
+    pub fn realloc(&mut self, new_len: usize) -> Result<(), &'static str> {
+        let new_mmap = if new_len == 0 {
+            None
+        } else {
+            if self.len() == 0 {
+                Some(Mmap::new(new_len)?)
+            } else {
+                let mut new_mmap = Mmap::new(new_len)?;
+
+                {
+                    let src = self
+                        .mmap
+                        .as_ref()
+                        .expect(
+                            "self.len() != 0;
+                            self.mmap is created if self.len() != 0;
+                            self.mmap is not `None`;
+                            qed",
+                        )
+                        .as_slice();
+                    let dst = new_mmap.as_slice_mut();
+                    dst[..src.len()].copy_from_slice(src);
+                }
+
+                Some(new_mmap)
+            }
+        };
+        self.mmap = new_mmap;
+        Ok(())
+    }
+
+    pub fn len(&self) -> usize {
+        self.mmap.as_ref().map(|m| m.len).unwrap_or(0)
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        self.mmap.as_ref().map(|m| m.as_slice()).unwrap_or(&[])
+    }
+
+    pub fn as_slice_mut(&mut self) -> &mut [u8] {
+        self.mmap
+            .as_mut()
+            .map(|m| m.as_slice_mut())
+            .unwrap_or(&mut [])
+    }
+
+    pub fn erase(&mut self) -> Result<(), &'static str> {
```

Comment:
```suggestion
    pub fn clear(&mut self) -> Result<(), &'static str> {
```

Reply:
I'd prefer to keep it `erase` since it conveys the purpose of this function better, as per discussion in https://github.com/paritytech/wasmi/pull/190#discussion_r300014756

Feel free to unresolve if you feel strongly about this